File: app/services/workspace/file_tools/pdf_tools.py
"""
PDFファイル用ツール

inspect_pdf_file: 構造確認（ページ数、目次）
read_pdf_pages: テキスト抽出
convert_pdf_to_images: ページを画像化してワークスペースに保存
"""

import io
from pathlib import Path
from typing import TYPE_CHECKING, Any

import structlog

if TYPE_CHECKING:
    from app.services.workspace_service import WorkspaceService

logger = structlog.get_logger(__name__)


async def inspect_pdf_file_handler(
    workspace_service: "WorkspaceService",
    tenant_id: str,
    conversation_id: str,
    args: dict[str, Any],
) -> dict[str, Any]:
    """
    PDFファイルの構造を確認

    Args:
        args:
            file_path: ファイルパス
    """
    file_path = args.get("file_path", "")

    try:
        import pymupdf
    except ImportError:
        return {
            "content": [{"type": "text", "text": "エラー: PyMuPDFライブラリがインストールされていません。"}],
            "is_error": True,
        }

    try:
        content, filename, content_type = await workspace_service.download_file(
            tenant_id, conversation_id, file_path
        )

        doc = pymupdf.open(stream=content, filetype="pdf")

        result_lines = [
            f"# PDF構造: {filename}",
            f"ファイルサイズ: {len(content) / 1024:.1f} KB",
            f"ページ数: {len(doc)}",
            "",
        ]

        # メタデータ
        metadata = doc.metadata
        if metadata:
            result_lines.append("## メタデータ")
            if metadata.get("title"):
                result_lines.append(f"- タイトル: {metadata['title']}")
            if metadata.get("author"):
                result_lines.append(f"- 作成者: {metadata['author']}")
            if metadata.get("subject"):
                result_lines.append(f"- 件名: {metadata['subject']}")
            result_lines.append("")

        # 目次（アウトライン）
        toc = doc.get_toc()
        if toc:
            result_lines.append("## 目次")
            for level, title, page in toc[:30]:  # 最大30項目
                indent = "  " * (level - 1)
                result_lines.append(f"{indent}- {title} (p.{page})")
            if len(toc) > 30:
                result_lines.append(f"  ... 他 {len(toc) - 30} 項目")
            result_lines.append("")

        # 各ページの概要
        result_lines.append("## ページ概要")
        for i, page in enumerate(doc):
            if i >= 10:  # 最大10ページまで
                result_lines.append(f"... 他 {len(doc) - 10} ページ")
                break

            page_num = i + 1
            text = page.get_text()
            text_length = len(text)

            # ページの種類を推定
            images = page.get_images()
            page_type = "テキスト主体"
            if len(images) > 0 and text_length < 200:
                page_type = "図表主体"
            elif len(images) > 0:
                page_type = "テキスト+図表"

            # テキストのプレビュー
            preview = text[:100].replace("\n", " ").strip()
            if len(text) > 100:
                preview += "..."

            result_lines.append(f"\n### ページ {page_num}")
            result_lines.append(f"- 種類: {page_type}")
            result_lines.append(f"- 文字数: {text_length}")
            result_lines.append(f"- 画像数: {len(images)}")
            if preview:
                result_lines.append(f"- プレビュー: {preview}")

        doc.close()

        result_text = "\n".join(result_lines)
        result_text += "\n\n---\n"
        result_text += "テキストを取得するには `read_pdf_pages` を使用してください。\n"
        result_text += "図表を確認するには `convert_pdf_to_images` で画像化してください。"

        return {
            "content": [{"type": "text", "text": result_text}],
        }
    except FileNotFoundError:
        return {
            "content": [{"type": "text", "text": f"ファイルが見つかりません: {file_path}"}],
            "is_error": True,
        }
    except Exception as e:
        logger.error("PDF構造確認エラー", error=str(e), file_path=file_path)
        return {
            "content": [{"type": "text", "text": f"読み込みエラー: {str(e)}"}],
            "is_error": True,
        }


async def read_pdf_pages_handler(
    workspace_service: "WorkspaceService",
    tenant_id: str,
    conversation_id: str,
    args: dict[str, Any],
) -> dict[str, Any]:
    """
    PDFページのテキストを抽出

    Args:
        args:
            file_path: ファイルパス
            pages: ページ指定（例: "1-5" または "1,3,5"）
    """
    file_path = args.get("file_path", "")
    pages_spec = args.get("pages", "1-10")

    try:
        import pymupdf
    except ImportError:
        return {
            "content": [{"type": "text", "text": "エラー: PyMuPDFライブラリがインストールされていません。"}],
            "is_error": True,
        }

    try:
        content, filename, content_type = await workspace_service.download_file(
            tenant_id, conversation_id, file_path
        )

        doc = pymupdf.open(stream=content, filetype="pdf")
        total_pages = len(doc)

        # ページ番号を解析
        page_numbers = _parse_pages(pages_spec, total_pages)

        result_lines = [
            f"# {filename} のテキスト",
            f"取得ページ: {pages_spec} (全{total_pages}ページ中)",
            "",
        ]

        for page_num in page_numbers:
            if page_num < 1 or page_num > total_pages:
                continue

            page = doc[page_num - 1]  # 0-indexed
            text = page.get_text()

            result_lines.append(f"## ページ {page_num}")
            result_lines.append("")

            if text.strip():
                result_lines.append(text.strip())
            else:
                # テキストがない場合
                images = page.get_images()
                if images:
                    result_lines.append("[このページは図表主体です。テキストは抽出できませんでした。]")
                    result_lines.append(f"[画像数: {len(images)}]")
                    result_lines.append("[内容を確認するには `convert_pdf_to_images` で画像化してください。]")
                else:
                    result_lines.append("[このページにテキストは含まれていません。]")

            result_lines.append("")
            result_lines.append("---")
            result_lines.append("")

        doc.close()

        result_text = "\n".join(result_lines)

        return {
            "content": [{"type": "text", "text": result_text}],
        }
    except FileNotFoundError:
        return {
            "content": [{"type": "text", "text": f"ファイルが見つかりません: {file_path}"}],
            "is_error": True,
        }
    except Exception as e:
        logger.error("PDFテキスト抽出エラー", error=str(e), file_path=file_path)
        return {
            "content": [{"type": "text", "text": f"読み込みエラー: {str(e)}"}],
            "is_error": True,
        }


async def convert_pdf_to_images_handler(
    workspace_service: "WorkspaceService",
    tenant_id: str,
    conversation_id: str,
    args: dict[str, Any],
) -> dict[str, Any]:
    """
    PDFページを画像に変換してワークスペースに保存

    Args:
        args:
            file_path: ファイルパス
            pages: ページ指定（例: "1-3" または "1,3,5"）最大5ページ
            dpi: 解像度（デフォルト: 150）
    """
    file_path = args.get("file_path", "")
    pages_spec = args.get("pages", "1")
    dpi = args.get("dpi", 150)

    try:
        import pymupdf
    except ImportError:
        return {
            "content": [{"type": "text", "text": "エラー: PyMuPDFライブラリがインストールされていません。"}],
            "is_error": True,
        }

    try:
        content, filename, content_type = await workspace_service.download_file(
            tenant_id, conversation_id, file_path
        )

        doc = pymupdf.open(stream=content, filetype="pdf")
        total_pages = len(doc)

        # ページ番号を解析
        page_numbers = _parse_pages(pages_spec, total_pages)

        # 最大5ページに制限
        if len(page_numbers) > 5:
            page_numbers = page_numbers[:5]
            logger.warning("画像変換を5ページに制限", requested=pages_spec)

        # ベースファイル名
        base_name = Path(filename).stem

        saved_paths = []
        zoom = dpi / 72  # 72 DPIが標準

        for page_num in page_numbers:
            if page_num < 1 or page_num > total_pages:
                continue

            page = doc[page_num - 1]

            # 画像に変換
            mat = pymupdf.Matrix(zoom, zoom)
            pix = page.get_pixmap(matrix=mat)

            # PNG形式でバイトに変換
            img_bytes = pix.tobytes("png")

            # 保存パス
            output_path = f"generated/{base_name}_page_{page_num}.png"

            # ワークスペースに保存（S3にアップロード）
            await workspace_service.s3.upload(
                tenant_id,
                conversation_id,
                output_path,
                img_bytes,
                "image/png",
            )

            # DBに登録
            await workspace_service.register_ai_file(
                tenant_id,
                conversation_id,
                output_path,
                is_presented=False,
            )

            saved_paths.append(output_path)
            logger.info("PDF→画像変換完了", page=page_num, path=output_path)

        doc.close()

        result_lines = [
            f"# PDF画像変換完了",
            f"ファイル: {filename}",
            f"変換ページ: {len(saved_paths)}ページ",
            "",
            "## 保存された画像",
        ]

        for path in saved_paths:
            result_lines.append(f"- {path}")

        result_lines.append("")
        result_lines.append("---")
        result_lines.append("画像を確認するには `read_image_file` でパスを指定してください。")

        return {
            "content": [{"type": "text", "text": "\n".join(result_lines)}],
        }
    except FileNotFoundError:
        return {
            "content": [{"type": "text", "text": f"ファイルが見つかりません: {file_path}"}],
            "is_error": True,
        }
    except Exception as e:
        logger.error("PDF画像変換エラー", error=str(e), file_path=file_path)
        return {
            "content": [{"type": "text", "text": f"変換エラー: {str(e)}"}],
            "is_error": True,
        }


def _parse_pages(pages_spec: str, total_pages: int) -> list[int]:
    """
    ページ指定を解析

    Args:
        pages_spec: "1-5" または "1,3,5" 形式
        total_pages: 総ページ数

    Returns:
        ページ番号のリスト（1始まり）
    """
    page_numbers = []

    # カンマ区切り
    parts = pages_spec.replace(" ", "").split(",")

    for part in parts:
        if "-" in part:
            # 範囲指定 (1-5)
            try:
                start, end = part.split("-")
                start_num = int(start)
                end_num = int(end)
                page_numbers.extend(range(start_num, min(end_num + 1, total_pages + 1)))
            except ValueError:
                continue
        else:
            # 単一ページ
            try:
                page_numbers.append(int(part))
            except ValueError:
                continue

    return sorted(set(page_numbers))

File: workspace-agent/src/file-tools/pdf-tools.ts
/**
 * PDF file tools
 *
 * inspect_pdf_file: Structure inspection (page count, TOC)
 * read_pdf_pages: Text extraction
 * convert_pdf_to_images: Convert pages to images (not supported in Node.js)
 */

import { getDocument, type PDFDocumentProxy } from "pdfjs-dist/legacy/build/pdf.mjs";
import { createLogger } from "../logger.js";
import {
  localFileToolHandler,
  formatToolSuccess,
  formatToolError,
  type ToolResult,
} from "./utils.js";

const logger = createLogger("pdf-tools");

// =============================================================================
// Internal Utilities
// =============================================================================

/**
 * Parse page specification string.
 * Supports "1-5" range and "1,3,5" comma-separated formats.
 * Returns sorted unique page numbers (1-based).
 */
function parsePages(pagesSpec: string, totalPages: number): number[] {
  const pageNumbers: number[] = [];
  const parts = pagesSpec.replace(/\s/g, "").split(",");

  for (const part of parts) {
    if (part.includes("-")) {
      try {
        const [startStr, endStr] = part.split("-");
        const startNum = parseInt(startStr, 10);
        const endNum = parseInt(endStr, 10);
        if (isNaN(startNum) || isNaN(endNum)) continue;
        const clampedEnd = Math.min(endNum, totalPages);
        for (let i = startNum; i <= clampedEnd; i++) {
          pageNumbers.push(i);
        }
      } catch {
        continue;
      }
    } else {
      const num = parseInt(part, 10);
      if (!isNaN(num)) {
        pageNumbers.push(num);
      }
    }
  }

  // Sort and deduplicate
  return [...new Set(pageNumbers)].sort((a, b) => a - b);
}

/**
 * Load a PDF document from a Buffer using pdfjs-dist.
 */
async function loadPdf(content: Buffer): Promise<PDFDocumentProxy> {
  const data = new Uint8Array(content);
  return await getDocument({ data, useSystemFonts: true }).promise;
}

/**
 * Extract text from a single PDF page.
 */
async function extractPageText(
  doc: PDFDocumentProxy,
  pageNum: number,
): Promise<string> {
  const page = await doc.getPage(pageNum);
  const textContent = await page.getTextContent();
  const lines: string[] = [];
  let lastY: number | null = null;

  for (const item of textContent.items) {
    if ("str" in item) {
      const y = (item as { transform: number[] }).transform[5];
      if (lastY !== null && Math.abs(y - lastY) > 2) {
        lines.push("\n");
      }
      lines.push(item.str);
      lastY = y;
    }
  }

  return lines.join("");
}

// =============================================================================
// Tool Handlers
// =============================================================================

/**
 * inspect_pdf_file handler - Inspect PDF structure
 */
export const inspectPdfFileHandler = localFileToolHandler(
  { logPrefix: "PDF構造確認" },
  async ({ content, filename }): Promise<ToolResult> => {
    const doc = await loadPdf(content);
    const numPages = doc.numPages;

    const resultLines: string[] = [
      `# PDF構造: ${filename}`,
      `ファイルサイズ: ${(content.length / 1024).toFixed(1)} KB`,
      `ページ数: ${numPages}`,
      "",
    ];

    // Metadata
    try {
      const metadata = await doc.getMetadata();
      const info = metadata?.info as Record<string, unknown> | undefined;
      if (info) {
        resultLines.push("## メタデータ");
        if (info.Title) {
          resultLines.push(`- タイトル: ${info.Title}`);
        }
        if (info.Author) {
          resultLines.push(`- 作成者: ${info.Author}`);
        }
        if (info.Subject) {
          resultLines.push(`- 件名: ${info.Subject}`);
        }
        resultLines.push("");
      }
    } catch {
      // Metadata not available
    }

    // Table of Contents (outline)
    try {
      const outline = await doc.getOutline();
      if (outline && outline.length > 0) {
        resultLines.push("## 目次");
        const maxItems = 30;
        let count = 0;

        const processOutline = (
          items: Array<{ title: string; items?: unknown[]; dest?: unknown }>,
          level: number,
        ) => {
          for (const item of items) {
            if (count >= maxItems) break;
            const indent = "  ".repeat(level);
            resultLines.push(`${indent}- ${item.title}`);
            count++;
            if (
              item.items &&
              Array.isArray(item.items) &&
              item.items.length > 0
            ) {
              processOutline(
                item.items as Array<{
                  title: string;
                  items?: unknown[];
                  dest?: unknown;
                }>,
                level + 1,
              );
            }
          }
        };

        processOutline(outline as Array<{ title: string; items?: unknown[]; dest?: unknown }>, 0);

        if (outline.length > maxItems) {
          resultLines.push(`  ... 他 ${outline.length - maxItems} 項目`);
        }
        resultLines.push("");
      }
    } catch {
      // Outline not available
    }

    // Page summaries (first 10 pages)
    resultLines.push("## ページ概要");
    const maxPreviewPages = Math.min(10, numPages);

    for (let i = 1; i <= maxPreviewPages; i++) {
      try {
        const text = await extractPageText(doc, i);
        const textLength = text.length;

        // Estimate page type
        let pageType = "テキスト主体";
        if (textLength < 200) {
          pageType = "図表主体";
        }

        // Text preview
        let preview = text.slice(0, 100).replace(/\n/g, " ").trim();
        if (text.length > 100) {
          preview += "...";
        }

        resultLines.push(`\n### ページ ${i}`);
        resultLines.push(`- 種類: ${pageType}`);
        resultLines.push(`- 文字数: ${textLength}`);
        if (preview) {
          resultLines.push(`- プレビュー: ${preview}`);
        }
      } catch {
        resultLines.push(`\n### ページ ${i}`);
        resultLines.push("- [テキスト抽出エラー]");
      }
    }

    if (numPages > 10) {
      resultLines.push(`... 他 ${numPages - 10} ページ`);
    }

    await doc.destroy();

    let resultText = resultLines.join("\n");
    resultText += "\n\n---\n";
    resultText +=
      "テキストを取得するには `read_pdf_pages` を使用してください。\n";
    resultText +=
      "図表を確認するには `convert_pdf_to_images` で画像化してください。";

    return formatToolSuccess(resultText);
  },
);

/**
 * read_pdf_pages handler - Extract text from PDF pages
 */
export const readPdfPagesHandler = localFileToolHandler(
  { logPrefix: "PDFテキスト抽出" },
  async ({ content, filename, args }): Promise<ToolResult> => {
    const pagesSpec = (args.pages as string) || "1-10";

    const doc = await loadPdf(content);
    const totalPages = doc.numPages;

    // Parse page numbers
    const pageNumbers = parsePages(pagesSpec, totalPages);

    const resultLines: string[] = [
      `# ${filename} のテキスト`,
      `取得ページ: ${pagesSpec} (全${totalPages}ページ中)`,
      "",
    ];

    for (const pageNum of pageNumbers) {
      if (pageNum < 1 || pageNum > totalPages) {
        continue;
      }

      resultLines.push(`## ページ ${pageNum}`);
      resultLines.push("");

      try {
        const text = await extractPageText(doc, pageNum);

        if (text.trim()) {
          resultLines.push(text.trim());
        } else {
          resultLines.push(
            "[このページは図表主体です。テキストは抽出できませんでした。]",
          );
          resultLines.push(
            "[内容を確認するには `convert_pdf_to_images` で画像化してください。]",
          );
        }
      } catch {
        resultLines.push("[このページのテキスト抽出に失敗しました。]");
      }

      resultLines.push("");
      resultLines.push("---");
      resultLines.push("");
    }

    await doc.destroy();

    return formatToolSuccess(resultLines.join("\n"));
  },
);

/**
 * convert_pdf_to_images handler - Convert PDF pages to images
 * Note: PDF to image conversion is not supported in Node.js without canvas.
 */
export const convertPdfToImagesHandler = localFileToolHandler(
  { logPrefix: "PDF画像変換" },
  async ({ content, filename, args }): Promise<ToolResult> => {
    const pagesSpec = (args.pages as string) || "1";

    // Load PDF to validate and get page count
    const doc = await loadPdf(content);
    const totalPages = doc.numPages;
    const pageNumbers = parsePages(pagesSpec, totalPages);
    await doc.destroy();

    // Limit to 5 pages
    const limitedPages = pageNumbers.slice(0, 5);
    if (pageNumbers.length > 5) {
      logger.warn(
        { requested: pagesSpec },
        "画像変換を5ページに制限",
      );
    }

    return formatToolError(
      `PDF to image conversion is not supported in this environment.\n\n` +
        `ファイル: ${filename}\n` +
        `リクエストページ: ${limitedPages.join(", ")} (全${totalPages}ページ)\n\n` +
        `代替手段:\n` +
        `- テキスト抽出には \`read_pdf_pages\` を使用してください。\n` +
        `- 図表の内容はテキストベースで確認してください。`,
    );
  },
);
